source/opt/propagator.h

```diff
@@ -0,0 +1,248 @@
+
+namespace spvtools {
+namespace opt {
+
+// Represents a CFG control edge.
+struct Edge {
+  explicit Edge(ir::BasicBlock* b1, ir::BasicBlock* b2)
+      : source(b1), dest(b2) {}
+  ir::BasicBlock* source;
+  ir::BasicBlock* dest;
+  bool operator<(const Edge& o) const {
+    return source < o.source || dest < o.dest;
+  }
+};
+
+// This class implements a generic value propagation algorithm based on the
+// conditional constant propagation algorithm proposed in
+//
+//      Constant propagation with conditional branches,
+//      Wegman and Zadeck, ACM TOPLAS 13(2):181-210.
+//
+//      A Propagation Engine for GCC
+//      Diego Novillo, GCC Summit 2005
+//      http://ols.fedoraproject.org/GCC/Reprints-2005/novillo-Reprint.pdf
+//
+// The purpose of this implementation is to act as a common framework for any
+// transformation that needs to propagate values from statements producing new
+// values to statements using those values.  Simulation proceeds as follows:
+//
+// 1- Initially, all edges of the CFG are marked not executable and the CFG
+//    worklist is seeded with all the statements in the entry basic block.
+//
+// 2- Every instruction I is simulated by calling a pass-provided function
+//    |visit_fn|.  This evaluation may produce 3 results:
+//
+//         SSAPropagator::kNotInteresting: Instruction I produces nothing of
+//             interest and does not affect any of the work lists.
+//
+//         SSAPropagator::kVarying: The value produced by I cannot be determined
+//             at compile time.  Further simulation of I is not required.
+//
+//         SSAPropagator::kInteresting: Instruction I produces a value that can
+//             be computed at compile time.
+//
+//    In addition to the return value, the |visit_fn| callback is responsible
+//    for actually saving the mapping between SSA ids and their values.  When an
+//    instruction I produces an interesting value, |visit_fn| should store that
+//    value into a value table.  At the end of propagation, values in the value
+//    table are guaranteed to be stable and can be replaced in the IR.
+//
+// 3- The propagator keeps two work queues.  Instructions are only added to
+//    these queues if they produce an interesting or varying value.
+//
+//      CFG blocks: contains the queue of blocks to be simulated.
+//             Blocks are added to this queue if their incoming edges are
+//             executable.
+//
+//      SSA Edges: An SSA edge is a def-use edge between a value-producing
+//              instruction and its use instruction.  The SSA edges list
+//              contains the statements at the end of a def-use edge that need
+//              to be re-visited when an instruction produces a kVarying or
+//              kInteresting result.
+//
+// 4- Simulation terminates when all work queues are drained.
+//
+//
+// EXAMPLE: Basic constant store propagator.
+//
+// Suppose we want to propagate all constant assignments of the form "OpStore
```

Comment:
I don't know if this is a good example.  No propagating seems to be going on, and it is unclear what should happen if there are two store to the same variable with different constants.  If I was trying to figure out how my |value_fn| function is suppose to work so I can use the propagator, I would be confused.